src/query/executor/request.go

```diff
@@ -0,0 +1,171 @@
+	switch s {
+	case compiling:
+		return "compiling"
+	case planning:
+		return "planning"
+	case executing:
+		return "executing"
+	case all:
+		return "all"
+	default:
+		return "unknown"
+	}
+}
+
+func (s State) durationString() string {
+	return fmt.Sprint("%s_duration_seconds", s)
+}
+
+// Request represents a single request.
+type Request struct {
+	engine     *Engine
+	params     models.RequestParams
+	parentSpan *span
+}
+
+func newRequest(ctx context.Context, engine *Engine, params models.RequestParams) (*Request, context.Context) {
+	parentSpan, parentCtx := startSpan(ctx, all.String(), engine.metrics.activeHist, engine.metrics.all)
+	r := &Request{engine: engine, params: params, parentSpan: parentSpan}
+	return r, parentCtx
+
+}
+
+func (r *Request) compile(ctx context.Context, parser parser.Parser) (parser.Nodes, parser.Edges, error) {
+	sp, _ := startSpan(ctx, compiling.String(), r.engine.metrics.compilingHist, r.engine.metrics.compiling)
+	// TODO: Change DAG interface to take in a context
+	nodes, edges, err := parser.DAG()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if r.params.Debug {
+		logging.WithContext(ctx).Info("compiled dag", zap.Any("nodes", nodes), zap.Any("edges", edges))
+	}
+
+	sp.finish()
+	return nodes, edges, nil
+}
+
+func (r *Request) plan(ctx context.Context, nodes parser.Nodes, edges parser.Edges) (plan.PhysicalPlan, error) {
+	sp, _ := startSpan(ctx, planning.String(), r.engine.metrics.planningHist, r.engine.metrics.planning)
+	lp, err := plan.NewLogicalPlan(nodes, edges)
+	if err != nil {
+		return plan.PhysicalPlan{}, err
+	}
+
+	if r.params.Debug {
+		logging.WithContext(ctx).Info("logical plan", zap.String("plan", lp.String()))
+	}
+
+	pp, err := plan.NewPhysicalPlan(lp, r.engine.store, r.params)
+	if err != nil {
+		return plan.PhysicalPlan{}, err
+	}
+
+	if r.params.Debug {
+		logging.WithContext(ctx).Info("physical plan", zap.String("plan", pp.String()))
+	}
+
+	sp.finish()
+	return pp, nil
+}
+
+func (r *Request) execute(ctx context.Context, pp plan.PhysicalPlan) (*ExecutionState, error) {
+	sp, _ := startSpan(ctx, executing.String(), r.engine.metrics.executingHist, r.engine.metrics.executing)
+	state, err := GenerateExecutionState(pp, r.engine.store)
+	// free up resources
+	if err != nil {
+		return nil, err
+	}
+
+	if r.params.Debug {
+		logging.WithContext(ctx).Info("execution state", zap.String("state", state.String()))
+	}
+
+	sp.finish()
+	return state, nil
+}
+
+func (r *Request) finish() {
+	r.parentSpan.finish()
+}
+
+// span is a simple wrapper around opentracing.Span in order to
+// get access to the duration of the span for metrics reporting.
+type span struct {
+	s            opentracing.Span
+	start        time.Time
+	duration     time.Duration
+	durationHist tally.Histogram
+	counter      *counterWithDecrement
+}
+
+func startSpan(ctx context.Context, operationName string, durationHist tally.Histogram, counter *counterWithDecrement) (*span, context.Context) {
```

Comment:
Want to maybe get one of the Jaeger guys to take a quick look at the diff? Couldn't hurt..